src/core/RecyclerListView.tsx

```diff
@@ -367,9 +370,19 @@ export default class RecyclerListView<P extends RecyclerListViewProps, S extends
         this._params.itemCount = newProps.dataProvider.getSize();
         this._virtualRenderer.setParamsAndDimensions(this._params, this._layout);
         this._virtualRenderer.setLayoutProvider(newProps.layoutProvider);
-        if (newProps.dataProvider.hasStableIds() && this.props.dataProvider !== newProps.dataProvider && newProps.dataProvider.requiresDataChangeHandling()) {
```

Comment:
I created a small snippet to verify that this change is necessary.

On this line 370, `requiresDataChangeHandling()` value returns false (but should be true) when:
```
new DataProvider((a, b) => a !== b)
  .cloneWithRows([1, 2])
  .cloneWithRows([1, 2, 3])
  .requiresDataChangeHandling() // returns false
```
or
```
new DataProvider((a, b) => a !== b)
    .cloneWithRows([])
    .cloneWithRows([1, 2, 3])
    .requiresDataChangeHandling() // returns false
```
(where first clone is the initial data provider and second clone is the next data provider)

In both instances, `handleDataSetChange` it not called and virtual renderer will continue using stable id method from the previous data provider, and so `getStableId(2)` would be out of bounds (when you scroll the list around).

This potentially means another patch is required inside `cloneWithRows` implementation, however regardless of where the bug originated from, I agree that updating stableId method is necessary always when data provider has changed.

Reply:
`requiresDataChangeHandling` is an optimization to avoid managing recycling keys in case of pagination where items are just getting added and no modification is there in existing ones.